Burst projectile retargeting

FS-21 opened the pull request:
- The firer will pick targets randomly.
- Works with missiles (no splits, airbusts, etc), cannons, lasers & spawners.
- A valid techno is required for trigger the logic.
- `OmniFire=yes` will make selectable any targets around the firer, limited by the weapon range.
- `OmniFire=no` will force the firer to pick targets in an area composed by the firer's weapon range around the original target intersected with the firer's weapon range around the firer.
- `RandomTarget.Spawners.MultipleTargets=true` gives each spawner it's own target.
- This logic should be used only in one weapon of the object.

[SOMEWEAPON]                                         ; WeaponType
RandomTarget=0.0                                           ; double or percentage
RandomTarget.Spawners.MultipleTargets=false ; boolean

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced `Burst.Retarget` allowing projectiles to retarget after each burst.
  - Added new projectile behavior for randomly selecting targets within range.
  - Customizable trajectory detonation and spawn distance adjustments for airstrike & spy plane.
  - Enabled application of `Verses` and `PercentAtMax` for negative damage.

- **Enhancements**
  - Improved targeting logic with the ability to refresh and update random targets for units.
  - Enhanced AI logic to handle new targeting and mission behaviors.

- **Documentation**
  - Updated `New-or-Enhanced-Logics.md` with examples and conditions for new projectile features.
  - Added summary of new features in `Whats-New.md`.

- **Refactor**
  - Centralized unit availability checks to `TechnoExt` class.
  - Removed redundant `IsUnitAvailable` function from `ScriptExt` class.

- **Bug Fixes**
  - Fixed issues with random target selection and assignment for various unit types.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

FS-21 committed:
First commit

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Added docs

FS-21 committed:
Removed comments

FS-21 commented:
Regarding how affects the Omnifire tag.
![imagen](https://github.com/Phobos-developers/Phobos/assets/1371539/cf4274fd-1e98-4594-9c48-6ea886a930c9)

mevitar commented:
Appears to work fine (even tried it for Trajectory=Straight projectiles), except for a case where a jumpjet without a turret and with OmniFire=no retargeted its burst fire into a landed AircraftType that was on a straight line behind. Same thing happened multuple times.

The jumpjet also had a secondary weapon that it switches to after running out of ammo, and that one has OmniFire=yes but it doesn't retarget.

FS-21 commented:
> Appears to work fine (even tried it for Trajectory=Straight projectiles), except for a case where a jumpjet without a turret and with OmniFire=no retargeted its burst fire into a landed AircraftType that was on a straight line behind. Same thing happened multuple times.
> 
> The jumpjet also had a secondary weapon that it switches to after running out of ammo, and that one has OmniFire=yes but it doesn't retarget.

What kind of weapons were used in the jumpjet case? I specified they should be of the type missile (but I noticed that invisible bullets works too).

mevitar commented:
I know you said that it was designed for regular missiles, but it also appears to works fine for Trajectory=Straight (also tried with Arcing=yes, and it doesn't work there at all).
I've encountered this problem by accident while testing Trajectory=Straight, but apparently it's not the projectile, because there's an issue with some jumpjets that they might attack an incorrect target, regardless of what weapon it has. Probably a combination of firing angle + the jumpjet trying to reposition itself and face the target is what caused such behavior, but i'm unable to reproduce it anymore.

FS-21 committed:
Tweak

MortonPL commented on src/Ext/Techno/Hooks.Firing.cpp:
```diff
@@ -794,3 +795,89 @@ DEFINE_HOOK(0x5223B3, InfantryClass_Approach_Target_DeployFireWeapon, 0x6)
+	if (!pWeaponExt || pWeaponExt->Burst_Retarget <= 0.0)
+		return 0;
+
+	int retargetProbability = pWeaponExt->Burst_Retarget > 1.0 ? 100 : (int)round(pWeaponExt->Burst_Retarget * 100);
```
```suggestion
	int retargetProbability = min((int)round(pWeaponExt->Burst_Retarget * 100), 100);
```

Perhaps a min() function would be more readable.

MortonPL commented on src/Ext/Techno/Hooks.Firing.cpp:
```diff
@@ -794,3 +795,89 @@ DEFINE_HOOK(0x5223B3, InfantryClass_Approach_Target_DeployFireWeapon, 0x6)
+		return 0;
+
+	int retargetProbability = pWeaponExt->Burst_Retarget > 1.0 ? 100 : (int)round(pWeaponExt->Burst_Retarget * 100);
+	int dice = ScenarioClass::Instance->Random.RandomRanged(1,100);
```
```suggestion
	int dice = ScenarioClass::Instance->Random.RandomRanged(1, 100);
```

MortonPL commented on src/Ext/Techno/Hooks.Firing.cpp:
```diff
@@ -794,3 +795,89 @@ DEFINE_HOOK(0x5223B3, InfantryClass_Approach_Target_DeployFireWeapon, 0x6)
+
+	for (auto pTarget : *TechnoClass::Array)
+	{
+		if (pTarget == pThis
+			|| pThisType->Immune
+			|| !EnumFunctions::IsTechnoEligible(pTarget, pWeaponExt->CanTarget, true)
+			|| (!pWeapon->Projectile->AA && pTarget->IsInAir())
+			|| (!pWeapon->Projectile->AG && !pTarget->IsInAir())
+			|| (!friendlyFire && (pThis->Owner->IsAlliedWith(pTarget) || ScriptExt:: IsUnitMindControlledFriendly(pThis->Owner, pTarget)))
+			|| pTarget->GetTechnoType()->Underwater && pTarget->GetTechnoType()->NavalTargeting == NavalTargetingType::Underwater_Never
+			|| pTarget->GetTechnoType()->Naval && pTarget->GetTechnoType()->NavalTargeting == NavalTargetingType::Naval_None)
+		{
+			continue;
+		}
```
Is there a better way to handle targeting other than manually checking all conditions? What about warhead versus, or CLEG'ed units, or shields, or cloaked units?

MortonPL reviewed (commented):
A few minor things and one question.

FS-21 committed:
v2.0 and tag rename

Now isn't required to have Burst > 1 in the weapon.

FS-21 commented:
I updated with the rewrite of the feature.
Now isn't required a Burst > 1 to work and it supports lasers and spawners (I don't know if there are more supported cases).

Now the tag is called "RandomTarget" (docs updated and the first post has been updated)
With the new tag _RandomTarget.Spawners.MultipleTargets=true_ the firer gives each spawner it's own random target.

FS-21 committed:
Small changes

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Point to latest YRpp

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Fixes

- Buidings retained target.
- Units retained targets when they were ordered to move.
- Small turret fix for non-laser weapons.

FS-21 committed:
tweak

FS-21 committed:
tweak

FS-21 committed:
Fix crash

FS-21 committed:
tweak

mevitar commented:
I found a potential issue with spawners on the Sept 22nd build, but can't test it if it still happens later because the newer build didn't build properly.

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

Metadorius commented on src/Ext/Techno/Body.cpp:
```diff
@@ -637,6 +639,213 @@ bool TechnoExt::IsTypeImmune(TechnoClass* pThis, TechnoClass* pSource)
+	for (auto pCandidate : *TechnoClass::Array)
+	{
+		if (pCandidate == pThis
+			|| !ScriptExt::IsUnitAvailable(pCandidate, true)
+			|| pThisType->Immune
+			|| !EnumFunctions::IsTechnoEligible(pCandidate, pWeaponExt->CanTarget, true)
+			|| (!pWeapon->Projectile->AA && pCandidate->IsInAir())
+			|| (!pWeapon->Projectile->AG && !pCandidate->IsInAir())
+			|| (!friendlyFire && (pThis->Owner->IsAlliedWith(pCandidate) || ScriptExt::IsUnitMindControlledFriendly(pThis->Owner, pCandidate)))
+			|| pCandidate->TemporalTargetingMe
+			|| pCandidate->BeingWarpedOut
+			|| (pCandidate->GetTechnoType()->Underwater && pCandidate->GetTechnoType()->NavalTargeting == NavalTargetingType::Underwater_Never)
+			|| (pCandidate->GetTechnoType()->Naval && pCandidate->GetTechnoType()->NavalTargeting == NavalTargetingType::Naval_None)
+			|| (pCandidate->CloakState == CloakState::Cloaked && !pThisType->Naval)
+			|| (pCandidate->InWhichLayer() == Layer::Underground))
+		{
+			continue;
+		}
+
+		int distanceFromAttacker = pThis->DistanceFrom(pCandidate);
+		if (distanceFromAttacker < minimumRange)
+			continue;
+
+		if (omniFire)
+		{
+			if (pCandidate->IsInAir())
+				range = airRange;
+
+			if (distanceFromAttacker <= range)
+				candidates.push_back(pCandidate);
+		}
+		else
+		{
+			int distanceFromOriginalTargetToCandidate = pCandidate->DistanceFrom(pThis->Target);
+			int distanceFromOriginalTarget = pThis->DistanceFrom(pThis->Target);
+
+			if (pCandidate->IsInAir())
+				range = airRange;
+
+			if (distanceFromAttacker <= range && distanceFromOriginalTargetToCandidate <= distanceFromOriginalTarget)
```
I am also not sure what does this function do and why is it in ScriptExt: `ScriptExt::IsUnitMindControlledFriendly(pThis->Owner, pCandidate)`

Starkku replied:
Function `ScriptExt::IsUnitMindControlledFriendly()` was originally placed in `ScriptExt` because that is the only place where it was called. If it is now being called elsewhere it should likely be relocated to `TechnoExt` since that would be more appropriate for function that checks `TechnoClass` state.

Metadorius replied:
We usually place static extension functions as if they were the non-static functions of the class that is being extended (think of C# extension methods), so if it accepts the TechnoClass, it should be in TechnoExt. I am not sure if the description you gave is accurate, I would say that it needs a clarification from @FS-21.

FS-21 replied:
Do you mean ScriptExt::IsUnitMindControlledFriendly()? I don't touch what other people do, like that ScriptExt::IsUnitAvailable().
I'll move them to TechnoExt then.

mevitar commented:
Something with weapon firing restrictions is terribly wrong, because GI-like infantry is be able to fire while their deploy sequence is playing. This is not the appropriate behavior, nor does it happen in the dev build or without Phobos - they should wait with firing their weapon until their deploy sequence finishes playing.

Retargetting doesn't work properly for Aircraft Carrier type spawners. Spawned aircrafts can get bugged, and, instead of following the proper target, they will try to fly outside of the map.
![retargetspawn2x](https://github.com/Phobos-developers/Phobos/assets/1313344/60a36d94-1951-4202-88d6-2959c61c8e6b)

FS-21 commented on docs/New-or-Enhanced-Logics.md:
````diff
@@ -1245,6 +1245,22 @@ ExtraWarheads=                  ; list of WarheadTypes
 ExtraWarheads.DamageOverrides=  ; list of integers
 ```
 
+### Projectile's random target
+- The firer will pick targets randomly.
+- Works with missiles (no splits, airbusts, etc), cannons, lasers & spawners.
+- A valid techno is required for trigger the logic.
+- `OmniFire=yes` will make selectable any targets around the firer, limited by the weapon range.
+- `OmniFire=no` will force the firer to pick targets in an area composed by the firer's weapon range around the original target intersected with the firer's weapon range around the firer.
+- `RandomTarget.Spawners.MultipleTargets=true` gives each spawner it's own target.
+- This logic should be used only in one weapon of the object.
+
+In `rulesmd.ini`:
+```ini
+[SOMEWEAPON]                                ; WeaponType
+RandomTarget=0.0                            ; double or percentage
+RandomTarget.Spawners.MultipleTargets=false ; boolean
+```
+
````
It was explained in that part, omit this documentation part.

FS-21 commented on src/Ext/Techno/Body.Update.cpp:
```diff
@@ -6,6 +6,7 @@
 #include <Ext/Anim/Body.h>
 #include <Ext/Bullet/Body.h>
 #include <Ext/House/Body.h>
+#include <Ext/Script/Body.h>
```
All headers here looks used in some functions so skip this suggestion.

FS-21 committed:
Moved IsUnitAvailable() in TechnoExt

FS-21 committed:
tweaks

FS-21 committed:
Changed name of GetRandomTarget()

With FindRandomTarget()

FS-21 commented on src/Ext/Techno/Body.Update.cpp:
```diff
@@ -719,3 +720,49 @@ void TechnoExt::UpdateSharedAmmo(TechnoClass* pThis)
+	if (!pExt)
+		return;
+
+	if (pThis->Target && pThis->SpawnManager && pExt->CurrentRandomTarget && ScriptExt::IsUnitAvailable(static_cast<TechnoClass*>(pExt->CurrentRandomTarget), true))
+	{
+		if (pThis->SpawnManager)
+		{
+			for (auto pSpawn : pThis->SpawnManager->SpawnedNodes)
+			{
+				if (!pSpawn->Unit)
+					continue;
+
+				auto pSpawnExt = TechnoExt::ExtMap.Find(pSpawn->Unit);
+				if (!pSpawnExt)
+					continue;
+
+				if (!pSpawnExt->CurrentRandomTarget)
+				{
+					pSpawnExt->CurrentRandomTarget = TechnoExt::GetRandomTarget(pThis);
+					pSpawn->Unit->Target = pSpawnExt->CurrentRandomTarget;
+				}
+				else if (pSpawn->Status == SpawnNodeStatus::Preparing && pSpawn->Unit->IsInAir())
+				{
+					if (!pSpawn->Unit->Target && pSpawnExt->CurrentRandomTarget)
+						pSpawn->Unit->Target = pSpawnExt->CurrentRandomTarget;
+				}
+			}
+		}
+	}
```
You are right!

FS-21 commented on src/Ext/Techno/Body.Update.cpp:
```diff
@@ -719,3 +720,49 @@ void TechnoExt::UpdateSharedAmmo(TechnoClass* pThis)
+		}
+	}
+
+	if (pExt->OriginalTarget && !pThis->Target && ScriptExt::IsUnitAvailable(static_cast<TechnoClass*>(pExt->OriginalTarget), true) && !pThis->IsInAir())
+	{
+		if (pExt->CurrentRandomTarget && ScriptExt::IsUnitAvailable(pExt->CurrentRandomTarget, true))
+			pThis->SetTarget(pExt->CurrentRandomTarget);
+		else
+			pThis->SetTarget(pExt->OriginalTarget);
+	}
```
IsUnitAvailable() uses 2 different parameter values so you are wrong in this case.

FS-21 commented:
> Something with weapon firing restrictions is terribly wrong, because GI-like infantry is be able to fire while their deploy sequence is playing. This is not the appropriate behavior, nor does it happen in the dev build or without Phobos - they should wait with firing their weapon until their deploy sequence finishes playing.

Then I need to add additional checks in that case.

> Retargetting doesn't work properly for Aircraft Carrier type spawners. Spawned aircrafts can get bugged, and, instead of following the proper target, they will try to fly outside of the map. ![retargetspawn2x](https://user-images.githubusercontent.com/1313344/282047376-60a36d94-1951-4202-88d6-2959c61c8e6b.gif) [ ![retargetspawn2x](https://user-images.githubusercontent.com/1313344/282047376-60a36d94-1951-4202-88d6-2959c61c8e6b.gif) ](https://user-images.githubusercontent.com/1313344/282047376-60a36d94-1951-4202-88d6-2959c61c8e6b.gif) [ ](https://user-images.githubusercontent.com/1313344/282047376-60a36d94-1951-4202-88d6-2959c61c8e6b.gif) This is with RandomTarget=100% and RandomTarget.Spawners.MultipleTargets=yes.

How is that Carrier weapon? The original Allied Carrier works fine with this:
```
[HornetLauncher]
Damage=1
ROF=150
Range=25
Spawner=yes
Projectile=Invisible
Speed=10
Warhead=Super
OmniFire=yes
RandomTarget=1.0 ; double or percentage
RandomTarget.Spawners.MultipleTargets=true ; boolean
```

FS-21 committed:
Tweaks based on feedback

mevitar commented:
It has nothing to do with the weapon.

The retargeting apparently ignores whether an object is actually targetable by the spawner weapon, which means it will still consider cloaked (and invisible) objects, objects with Immune=yes, and objects with 0% verses as legal targets. Thus, this behavior, with spawned aircraft flying away because it is unable to attack the thing it was ordered to target.

Don't know if this happens to non-spawner weapons.

FS-21 commented:
Yes, currently the method for picking the target doesn't ask for the spawned weapons capabilities so I'll address the deploy & this issue in the weekend.

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/burst-retarget

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/burst-retarget

FS-21 committed:
fixed weird merge

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Fix crashes

Rewriting some parts

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Fix crashes

mevitar commented:
> The retargeting apparently ignores whether an object is actually targetable by the spawner weapon, which means it will still consider cloaked (and invisible) objects, objects with Immune=yes, and objects with 0% verses armor as legal targets. Thus, this behavior, with spawned aircraft flying away because it is unable to attack the thing it was ordered to target.

I don't know if this was supposed to be fixed or not, but i got to test it in the newest build available, and the issue persists. Cloaked units still cause spawners to fly away.

FS-21 commented:
I forgot about that case x'D

When I arrive home I would like to update the code.

mevitar commented:
No need to hurry.

On a side note:
> Something with weapon firing restrictions is terribly wrong, because GI-like infantry is be able to fire while their deploy sequence is playing. This is not the appropriate behavior, nor does it happen in the dev build or without Phobos - they should wait with firing their weapon until their deploy sequence finishes playing.

There's something else going with this, it seems. I don't know how did i manage to make it not happen without Phobos before, but i retested it now and it does happen, and i didn't really change anything for the unit in question (at least i don't think i did). It appears to happen randomly, and the more infantry units are in a group, the more noticeable it becomes.

Either way, no need to fix anything with this one, I'll probably report that as a separate issue.

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Fix crash

Produced when clicked stop command in the middle of an attack.

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Moved code into new files

FS-21 committed:
Moving more code

And renaming some events

FS-21 committed:
Code rewrite

The major part of the code was rewritten.

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
small fix

FS-21 committed:
small doc fix

FS-21 committed:
Small changes

FS-21 committed:
tweak

FS-21 committed:
tweak

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

FS-21 committed:
Merge branch 'develop' into feature/burst-retarget

TaranDahl commented on docs/New-or-Enhanced-Logics.md:
````diff
@@ -2558,6 +2558,22 @@ ExtraWarheads.DetonationChances=  ; List of floating-point values (percentage or
 ```
 
+### Projectile's random target
+- The firer will pick targets randomly.
````
The title and the description are inconsistent.
Firer and proj are two completely different things. Your title says proj, when your description says firer.

TaranDahl reviewed (changes requested):
There are too many redundant checks, and I can't bear to look at them.
I imagine that the implementation of this logic should be very concise and doesn't require that much code. I think you're trying to implement the function you want in a completely wrong way, and according to feedback from others, it's not working the way you hoped.
Please rewrite the whole using the following logic, or delete unnecessary codes, briefly summarize the operational logic of your program in words and convince me why not to use the following logic:
For the case of ordinary projectiles: When a unit is about to fire, call `CanAutoTargetObject` and `GetFireError` on all units within range. If the former returns false, or the latter returns a code other than Facing and OK, exclude that unit. Randomly select one from the available units and fire at it (should hook into Fire func and change the FireError check).
For the case of the spawner: Declare a target list for SpawnerExt, and replace all reference to the vanilla property Target with the target list. When spawning spawners, find a valid target and add it to the target list. The spawners will act according to the target corresponding to themselves in the target list.

FS-21 commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -2558,6 +2558,22 @@ ExtraWarheads.DetonationChances=  ; List of floating-point values (percentage or
+- `OmniFire=yes` will make selectable any targets around the firer, limited by the weapon range.
+- `OmniFire=no` will force the firer to pick targets in an area composed by the firer's weapon range around the original target intersected with the firer's weapon range around the firer.
+- `RandomTarget.Spawners.MultipleTargets=true` gives each spawner it's own target.
+- This logic should be used only in one weapon of the object.
```
Why all weapons should be affected at the same time by this logic instead of per weapon?

FS-21 commented:
All the code has been rewritten.
Updated the first message of this PR.

TaranDahl commented on src/Ext/Techno/Body.RandomTarget.cpp:
```diff
@@ -0,0 +1,202 @@
+		}
+	}
+
+	for (const auto pCandidate : TechnoClass::Array)
```
Use the helper

TaranDahl commented on src/Ext/Techno/Body.RandomTarget.cpp:
```diff
@@ -0,0 +1,202 @@
+		return pOriginalTarget;
+
+	const int pick = ScenarioClass::Instance->Random.RandomRanged(0, static_cast<int>(candidates.size()) - 1);
+	return candidates[pick];
```
Why reinvent the wheel for GetFireError?

TaranDahl commented on src/Ext/WeaponType/Body.h:
```diff
@@ -113,6 +112,13 @@ class WeaponTypeExt final : public AbstractTypeExt
+	Valueable<bool> RandomTarget_Spawners_RememberTargets;
+	Valueable<double> RandomTarget_MissChance;
+	Valueable<double> RandomTarget_FriendlyFireChance;
+	ValueableVector<TechnoTypeClass*> OnlyTargetTechnos;
```
This is not in the doc.  For what reason?

FS-21 replied:
What tag?

TaranDahl commented on src/Ext/Techno/Body.RandomTarget.cpp:
```diff
@@ -0,0 +1,202 @@
+		// Air / Ground capabilities
+		if (pWeapon->Projectile)
+		{
+			const bool inAir = pCandidate->IsInAir();
+			if (!pWeapon->Projectile->AA && inAir)
+				continue;
+			if (!pWeapon->Projectile->AG && !inAir)
+				continue;
+		}
+
+		// House relationships
+		if (targetAllies)
+		{
+			const bool isFriendly = pFirer->Owner && (pFirer->Owner->IsAlliedWith(pCandidate->Owner) || pCandidate->Owner == pFirer->Owner);
+			if (!isFriendly)
+				continue;
+		}
+		else
+		{
+			const bool isSameHouseAsTarget = pOrigHouse && (pCandidate->Owner == pOrigHouse);
+			const bool isEnemy = pFirer->Owner && (!pFirer->Owner->IsAlliedWith(pCandidate->Owner) && !ScriptExt::IsMindControlledByEnemy(pFirer->Owner, pCandidate));
+
+			if (!isSameHouseAsTarget && !isEnemy)
+				continue;
+		}
+
+		// Phobos weapon eligibility filters
+		if (!EnumFunctions::IsTechnoEligible(pCandidate, pWeaponExt->CanTarget, true))
+			continue;
+
+		if (!pWeaponExt->CanOnlyTargetTheseTechnos(pCandidateType))
+			continue;
+
+		if (!pWeaponExt->IsHealthInThreshold(pCandidate))
+			continue;
+
+		if (!pWeaponExt->HasRequiredAttachedEffects(pCandidate, pFirer))
+			continue;
+
+		candidates.push_back(pCandidate);
```
Why reinvent the wheel for GetFireError?

FS-21 replied:
Reused old code from the first implementations.
